### PR TITLE
Allow app to set the C flag in the packet header for control message.

### LINKED ISSUE #4641: Add interface in Transport Layer to allows the messaging layer configure packet header
#### Problem
Currently, we only can configure payload header before sending message to transport layer. Packetheader are auto-configured in Transport layer 

 #### Proposed Solution
Add interface in Transport Layer to allows the messaging layer configure packet header

### PR DESCRIPTION
**Pull request by yufengwangca**
#### Problem
Currently, we only can configure payload header before sending message to transport layer. Packetheader are auto-configured in Transport layer,  for example, we shall set the C flag in the packet header for certain type of control message.

 #### Summary of Changes
Add interface in Transport Layer to allows the messaging layer configure packet header

 Fixes #4641

### CODE REVIEW

File: src/transport/SecureSessionMgr.h [hunk 1]
```diff
@@ -181,8 +181,8 @@ class DLL_EXPORT SecureSessionMgr : public TransportMgrDelegate
      *   to retransmit it, will be stored in *bufferRetainSlot.
      */
     CHIP_ERROR SendMessage(SecureSessionHandle session, System::PacketBufferHandle && msgBuf);
-    CHIP_ERROR SendMessage(SecureSessionHandle session, PayloadHeader & payloadHeader, System::PacketBufferHandle && msgBuf,
-                           EncryptedPacketBufferHandle * bufferRetainSlot = nullptr);
+    CHIP_ERROR SendMessage(SecureSessionHandle session, PayloadHeader & payloadHeader, PacketHeader & packetHeader,
```
Comment (andy31415) [inline src/transport/SecureSessionMgr.h:184]: Could you add a comment here about what each argument is for?

I am a bit worried that we start adding too many parameters here. 5 is already a lot.

Also worried about separation of concerns: I imagine we want objects to have a clear control. Should we instead set some sending flags? Passing in an entire packet header that is generally fully managed by the secure session (secure session defines lengths, encryptions and such) seems too much. We are sending an entire object to allow the caller to set a single bit flag.

Comment (yufengwangca) [reply src/transport/SecureSessionMgr.h:194]: Added, we already have the sending flags defined in the messaging layer to simply the configuration. 
https://github.com/project-chip/connectedhomeip/blob/master/src/messaging/Flags.h#L60

The clients of messaging layer only need to configure the SendFlags, PayloadHeader and PacketHeader are defined in transport layer which are aligned with spec. Currently, PacketHeader is always hard coded to default value, which is not exposed to upper layer to configure.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]
I agree with @andy31415.  Using `PacketHeader` as the "pass some flags" API seems really weird.  If consumers are never expected to interact with any of the other parts of `PacketHeader`, I would rather we passed the flags directly.

Separately from all this, though: are we expecting to have consumers send non-CHIP-defined control messages?  And if not, can't we just set the C flag based on the protocol and message type?

**Comment by yufengwangca** [comment]
> I agree with @andy31415. Using `PacketHeader` as the "pass some flags" API seems really weird. If consumers are never expected to interact with any of the other parts of `PacketHeader`, I would rather we passed the flags directly.
> 
> Separately from all this, though: are we expecting to have consumers send non-CHIP-defined control messages? And if not, can't we just set the C flag based on the protocol and message type?

@bzbarsky-apple The consumers should not talk to Transport layer directly, (We still have some legacy to cleanup), that is why we introduced SendFlags exactly for this reason. https://github.com/project-chip/connectedhomeip/blob/master/src/messaging/Flags.h#L60

We can move down SendFlags from messaging layer API to transport layer API, using SendFlags instead of PacketHeader and PayloadHeader. I remember I got comments before to isolate this to messaging layer and so that lower layer does not need to depend on the higher layer. I think this make sense to me. What do you think?

**Comment by bzbarsky-apple** [comment]
> The consumers should not talk to Transport layer directly

OK, let's take that as an assumption.

> we introduced SendFlags exactly for this reason

Right, that makes sense.

> We can move down SendFlags from messaging layer API to transport layer API, using SendFlags instead of PacketHeader and PayloadHeader

I'm not sure I follow.  If what we are trying to communicate are single bits, then SendFlags works.  If we are trying to communicate the sort of info that actually belongs in the two header types, then we need the actual header types....

> I remember I got comments before to isolate this to messaging layer so that lower layer does not depend on the higher layer. 

I understand that desire; I wonder to what extent these layering distinctions are artificial and limiting having consumer-facing APIs that are easy to use and hard to misuse.  I just filed https://github.com/project-chip/connectedhomeip/issues/4786 with some thoughts, that might be related, fwiw.

> Using protocol type & message type to hard code the C flag might restrict some flexibility

That is correct.  What I am asking is whether it restricts flexibility that the spec allows or not.  And for non-spec messages whether it restricts flexibility that we want to allow or not.

> Such as Status Messages (within the Common Protocol) is used return single operation status codes via a common pathway, which could set C flag or not depending on different scenarios.

I don't see anything in the spec that allows sending "Status Message" (which is actually not defined in the spec right now) with "C Flag" set.  The only spec-defined messages that currently set "C Flag" are "MsgCounterSyncReq" and "MsgCounterSyncRsp".

Again, I understand the desire to have a flexible API that allows all possible behaviors and rely on higher layers to not do the wrong thing, but I think we should strongly think about the ergonomics and consider nudging, if not forcing, the right thing to happen.

**Comment by yufengwangca** [comment]
@bzbarsky-apple 

Regarding "can we just set the C flag based on the protocol and message type?",  where is the best place to set this flag? One possible place is when we set message type and protocol type to PayloadHeader, but PayloadHeader is currently independent from PacketHeader, it can not set C flag there.

Another place is within SendMessage API of SecureSessionMgr, we sniff the incoming message and set the C Flag depending on the message type. 

Currently, we always use the default value of PacketHeader, which seems not right, this block all following features which need to configure the PacketHeader. 

We either use SendFlags in SendMessage of SecureSessionMgr (this was my original proposal) or just hide all of them within messaging layer as this PR.

**Comment by yufengwangca** [comment]
I will set C flag for  "MsgCounterSyncReq" and "MsgCounterSyncRsp" explicitly for now since only these two messages are using this flag. Reopen this ticket if we have more generic usage of C flag

**Comment by bzbarsky-apple** [comment]
My apologies for the laggy response here.  Partly, I did not have a great answer to the questions in https://github.com/project-chip/connectedhomeip/pull/4771#issuecomment-777929335 without doing a bunch of digging first....

I think if we can set the C flag automatically for the moment, by sniffing the incoming message type, we should do that.  If it turns out we need something more general later we can try to figure out how to do that.

**Pull request closed without merging**